x-pack/plugins/security_solution/public/management/components/console/components/console_manager/console_manager.tsx

```diff
@@ -0,0 +1,306 @@
+      }
+    },
+    [] // << IMPORTANT: this callback should have dependencies or only immutable dependencies
+  );
+
+  const getList = useCallback<ConsoleManagerClient['getList']>(<
+    Meta extends object = Record<string, unknown>
+  >() => {
+    return Object.values(consoleStorage).map(
+      (managedConsole) => managedConsole.client
+    ) as ReadonlyArray<Readonly<RegisteredConsoleClient<Meta>>>;
+  }, [consoleStorage]); // << This callack should always use `consoleStorage`
+
+  const isVisible = useCallback((id: string): boolean => {
+    if (consoleStorageRef.current?.[id]) {
+      return consoleStorageRef.current[id].isOpen;
+    }
+
+    return false;
+  }, []); // << IMPORTANT: this callback should have no dependencies
+
+  const register = useCallback<ConsoleManagerClient['register']>(
+    ({ id, title, meta, consoleProps, ...otherRegisterProps }) => {
+      if (consoleStorage[id]) {
+        throw new Error(`Console with id ${id} already registered`);
+      }
+
+      const managedKey = Symbol(id);
+      // Referencing/using the interface methods here (defined in the outer scope of this function)
+      // is ok because those are immutable and thus will not change between state changes
+      const showThisConsole = show.bind(null, id);
+      const hideThisConsole = hide.bind(null, id);
+      const terminateThisConsole = terminate.bind(null, id);
+      const isThisConsoleVisible = isVisible.bind(null, id);
+
+      const managedConsole: ManagedConsole = {
+        ...otherRegisterProps,
+        client: {
+          id,
+          title,
+          meta,
+          // The use of `setTimeout()` below is needed because this client interface can be consumed
+          // prior to the component state being updated. Placing a delay on the execution of these
+          // methods allows for state to be updated first and then the action is applied.
+          // So someone can do: `.register({...}).show()` and it will work
+          show: () => {
+            setTimeout(showThisConsole, 0);
+          },
+          hide: () => {
+            setTimeout(hideThisConsole, 0);
+          },
+          terminate: () => {
+            setTimeout(terminateThisConsole, 0);
+          },
+          isVisible: () => isThisConsoleVisible(),
+        },
+        consoleProps,
+        console: <Console {...consoleProps} managedKey={managedKey} key={id} />,
+        isOpen: false,
+        key: managedKey,
+      };
+
+      setConsoleStorage((prevState) => {
+        return {
+          ...prevState,
+          [id]: managedConsole,
+        };
+      });
+
+      return managedConsole.client;
+    },
+    [consoleStorage, hide, isVisible, show, terminate]
+  );
+
+  const consoleManagerClient = useMemo<ConsoleManagerClient>(() => {
+    return {
+      register,
+      show,
+      hide,
+      terminate,
+      getOne,
+      getList,
+    };
+  }, [getList, getOne, hide, register, show, terminate]);
```

Comment:
Do we really need to memoize this?  We're not doing any complex computations here.

Reply:
The memoization here is just to ensure that the object that holds the interface does not change between re-renders and that the `useConsoleManager()` hook will not cause downstream re-renders if the interface has not changed.